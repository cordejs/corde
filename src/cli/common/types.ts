import { Command } from "commander";

export type ParamsFrom = "args" | "options" | "both";

export interface ICommandOptions {
  flags: string;
  description?: string;
  defaultValue?: string;
}

export interface ICliCommandConstructor {
  name?: string;
  program: Command;
  paramsFrom?: ParamsFrom;
}

export interface ICliCommand {
  readonly paramsFrom?: ParamsFrom;
  readonly command: Command;
  handler(...args: any[]): Promise<void> | void;
}

export interface ICliCommandConstructable {
  new (command: Command): ICliCommand;
}
