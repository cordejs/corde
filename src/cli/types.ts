export interface ICommand {
  action(...args: any[]): Promise<void> | void;
}
