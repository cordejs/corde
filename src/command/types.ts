import { ICordeBot } from "../types";

export interface ICreateMatcherParam {
  matcher?: string;
  isNot?: boolean;
  commandName?: string | boolean | number;
  channelId?: string;
  guildId?: string;
  isDebug?: boolean;
  isCascade?: boolean;
  cordeBot?: ICordeBot;
  trace?: string;
  utility?: boolean;
  mustSendCommand?: boolean;
}
