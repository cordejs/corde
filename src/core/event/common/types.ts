import { Client, ClientEvents } from "discord.js";

export type EventNames = keyof ClientEvents;

export type Spread<T extends EventNames> = ClientEvents[T][1] extends undefined
  ? ClientEvents[T] extends (infer U)[]
    ? U
    : never
  : ClientEvents[T];

export interface IDiscordEvent<TEvent extends EventNames, TFilter> {
  canListen(): boolean;
  getIntents(): string[];
  on(fn: (...args: ClientEvents[TEvent]) => void): this;
  once(filter?: TFilter): Promise<Spread<TEvent>>;
}

export interface IDiscordEventConstructable<TEvent extends EventNames, TFilter> {
  new (client: Client): IDiscordEvent<TEvent, TFilter>;
}

export type IntentType =
  | "GUILDS"
  | "GUILD_MEMBERS"
  | "GUILD_BANS"
  | "GUILD_EMOJIS_AND_STICKERS"
  | "GUILD_INTEGRATIONS"
  | "GUILD_WEBHOOKS"
  | "GUILD_INVITES"
  | "GUILD_VOICE_STATES"
  | "GUILD_PRESENCES"
  | "GUILD_MESSAGES"
  | "GUILD_MESSAGE_REACTIONS"
  | "GUILD_MESSAGE_TYPING"
  | "DIRECT_MESSAGES"
  | "DIRECT_MESSAGE_REACTIONS"
  | "DIRECT_MESSAGE_TYPING"
  | "GUILD_SCHEDULED_EVENTS";

export type EventClass = {
  [k in IntentType]: {
    bitfield: number;
    events: EventNames[];
  };
};
