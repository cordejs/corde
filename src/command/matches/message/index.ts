export { ToAddReaction } from "./toAddReaction";
export { ToRemoveReaction } from "./toRemoveReaction";
export { ToReturn } from "./toReturn";
export { ToUnPinMessage } from "./toUnpinMessage";
export { ToPinMessage } from "./toPinMessage";
export { ToEditMessage } from "./toEditMessage";
export { ToMessageContentContains } from "./toMessageContentContains";
export { ToEmbedMatch } from "./toEmbedMatch";
