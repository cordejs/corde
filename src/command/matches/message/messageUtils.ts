import { EmbedFieldData, Message, MessageEmbed } from "discord.js";
import { diff } from "jest-diff";
import { MessageType, Primitive } from "../../../types";
import { deepEqual } from "../../../utils/deepEqual";
import { formatObject } from "../../../utils/formatObject";
import { isPrimitiveValue } from "../../../utils/isPrimitiveValue";
import { typeOf } from "../../../utils/typeOf";
import { CommandState } from "../CommandState";

export namespace messageUtils {
  export function validateExpect(matcher: CommandState, expect: Primitive | corde.IMessageEmbed) {
    if (!isPrimitiveValue(expect) && typeOf(expect) !== "object") {
      return matcher.createReport(
        "expected: expect value to be a primitive value (string, boolean, number) or an IMessageEmbed\n",
        `received: ${typeOf(expect)}`,
      );
    }
    return null;
  }

  export function createReportForExpectAndResponse(
    matcher: CommandState,
    expect: Primitive | corde.IMessageEmbed,
    returnedMessage: Message,
  ) {
    matcher.hasPassed = isMessagesEquals(returnedMessage, expect);
    matcher.invertHasPassedIfIsNot();

    if (matcher.hasPassed) {
      return matcher.createPassTest();
    }

    let embedReturned: corde.IMessageEmbed | undefined;
    if (returnedMessage?.embeds[0]) {
      embedReturned = getMessageByType(returnedMessage, "embed") as corde.IMessageEmbed;
    }

    if (typeOf(expect) === "object" && embedReturned) {
      return matcher.createReport(diff(embedReturned, expect));
    }

    if (typeOf(expect) === "object" && !embedReturned) {
      return matcher.createReport(
        `expected: ${formatObject(expect)}\n`,
        `received: '${returnedMessage?.content}'`,
      );
    }

    if (typeOf(expect) === "string" && embedReturned) {
      return matcher.createReport(
        `expected: '${expect}'\n`,
        `received: ${formatObject(embedReturned)}`,
      );
    }

    return matcher.createReport(
      `expected: '${expect}'\n`,
      `received: '${returnedMessage?.content}'`,
    );
  }

  export function isMessagesEquals(
    returnedMessage: Message,
    expectation: Primitive | corde.IMessageEmbed,
  ) {
    if (!returnedMessage || !expectation) {
      return false;
    }

    const embed = returnedMessage.embeds[0];
    if (isPrimitiveValue(expectation) && !embed) {
      return expectation == returnedMessage.content;
    }

    if (embed && typeOf(expectation) === "object") {
      const embedInternal = messageEmbedToMessageEmbedInterface(embed);
      return deepEqual(expectation, embedInternal);
    }

    return false;
  }

  /**
   * Format Discord responses
   *
   * @param answer Discord response for a message sent
   *
   * @param type Type expected of that message
   *
   * @description Discord adds some attributes that are not present in embed message before it is sent
   *
   *  This is data **before** send to Discord
   *
   *  ```javascript
   *   "image": {
   *       "url": "https://i.imgur.com/wSTFkRM.png"
   *   },
   *   "thumbnail": {
   *       "url": "https://i.imgur.com/wSTFkRM.png"
   *   }
   *  ```
   *
   *  And this is part of embed message **after** get from Discord
   *
   *  ```javascript
   *   "image": {
   *     "height": 0,
   *     "proxyURL": "https://images-ext-2.discordapp.net/external/DoAGN014Q46B7iDBr2VJyHUL59QLSWdEAZ5wOoWe8CY/https/i.imgur.com/wSTFkRM.png",
   *     "url": "https://i.imgur.com/wSTFkRM.png",
   *     "width": 0
   *   },
   *   "thumbnail": {
   *       "height": 0,
   *       "proxyURL": "https://images-ext-2.discordapp.net/external/DoAGN014Q46B7iDBr2VJyHUL59QLSWdEAZ5wOoWe8CY/https/i.imgur.com/wSTFkRM.png",
   *       "url": "https://i.imgur.com/wSTFkRM.png",
   *      "width": 0
   *  }
   *  ```
   */
  export function getMessageByType(answer: Message | MessageEmbed, type: MessageType) {
    if (type === "embed") {
      const embed = answer instanceof Message ? answer.embeds[0] : answer;
      if (!embed) {
        return null;
      }

      return messageEmbedToMessageEmbedInterface(embed as MessageEmbed);
    }

    if (answer instanceof Message) {
      return answer.content;
    }
    return answer;
  }

  export function humanizeMessageIdentifierObject(
    msgIdentifier?: corde.IMessageIdentifier | corde.IMessageEditedIdentifier,
  ) {
    if (!msgIdentifier) {
      return "";
    }
    if (msgIdentifier?.id) {
      return `message of id ${msgIdentifier.id}`;
    }
    if ((msgIdentifier as corde.IMessageIdentifier).content) {
      return `message of content "${(msgIdentifier as corde.IMessageIdentifier).content}"`;
    }
    if ((msgIdentifier as corde.IMessageEditedIdentifier).oldContent) {
      return `message of content "${(msgIdentifier as corde.IMessageEditedIdentifier).oldContent}"`;
    }
    return "";
  }

  export function messageEmbedToMessageEmbedInterface(message?: MessageEmbed) {
    if (!message) {
      return {};
    }

    const embedLike: corde.IMessageEmbed = {};

    if (message.url) {
      embedLike.url = message.url;
    }

    if (message.timestamp) {
      embedLike.timestamp = message.timestamp;
    }

    if (message.author) {
      if (message.author.iconURL || message.author.url) {
        embedLike.author = {
          iconURL: message.author.iconURL,
          name: message.author.name,
          url: message.author.url,
        };
      } else if (message.author.name) {
        embedLike.author = message.author.name;
      }
    }

    if (message.color) {
      embedLike.color = message.color;
    }

    if (message.description) {
      embedLike.description = message.description;
    }

    if (message.fields && message.fields.length) {
      embedLike.fields = [];
      message.fields.forEach((field) => {
        embedLike.fields?.push({
          name: field.name,
          value: field.value,
          inline: !!field.inline,
        });
      });
    }

    if (message.footer) {
      embedLike.footer = message.footer;
    }

    if (message.image) {
      if (message.image.height || message.image.width) {
        embedLike.image = {
          url: message.image.url,
          height: message.image.height,
          width: message.image.width,
        };
      } else if (message.image.url) {
        embedLike.image = message.image.url;
      }
    }

    if (message.title) {
      embedLike.title = message.title;
    }

    if (message.thumbnail) {
      embedLike.thumbnailUrl = message.thumbnail;
    }
    return embedLike;
  }

  export function embedMessageInterfaceToMessageEmbed(embedLike?: corde.IMessageEmbed) {
    const embed = new MessageEmbed();
    if (!embedLike || typeOf(embedLike) !== "object") {
      return embed;
    }

    if (embedLike.author) {
      if (typeof embedLike.author === "string") {
        embed.setAuthor({
          name: embedLike.author,
        });
      } else {
        embed.setAuthor({
          name: embedLike.author.name,
          iconURL: embedLike.author.iconURL,
          url: embedLike.author.url,
        });
      }
    }

    if (embedLike.color) {
      embed.setColor(embedLike.color);
    }

    if (embedLike.description) {
      embed.setDescription(embedLike.description);
    }

    if (embedLike.fields) {
      embed.addFields(
        ...embedLike.fields.map<EmbedFieldData>((field) => {
          return {
            name: field.name,
            value: field.value,
            inline: !!field.inline,
          };
        }),
      );
    }

    if (embedLike.footer) {
      embed.setFooter(embedLike.footer);
    }

    if (embedLike.image) {
      if (typeof embedLike.image === "string") {
        embed.setImage(embedLike.image);
      } else {
        embed.setImage(embedLike.image.url);
      }
    }

    if (embedLike.thumbnailUrl) {
      embed.thumbnail = embedLike.thumbnailUrl;
    }

    if (embedLike.timestamp) {
      embed.setTimestamp(embedLike.timestamp);
    }

    if (embedLike.title) {
      embed.setTitle(embedLike.title);
    }

    if (embedLike.url) {
      embed.setURL(embedLike.url);
    }

    return embed;
  }
}
