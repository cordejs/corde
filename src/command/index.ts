/* eslint-disable no-console */
import { forEachPropsName } from "../utils/forEachPropsName";
import { getStackTrace } from "../utils/getStackTrace";
import { isNullOrUndefined } from "../utils/isNullOrUndefined";
import { any } from "../expect/asymmetricMatcher";
import * as matchers from "./matches";
import runtime from "../core/runtime";
import { ICordeBot, ITestReport } from "../types";
import { CommandState } from "./matches/CommandState";

interface ICreateMatcherParam {
  matcher: string;
  isNot: boolean;
  commandName?: string | boolean | number;
  channelId?: string;
  guildId?: string;
  isDebug: boolean;
  isCascade?: boolean;
  cordeBot?: ICordeBot;
  trace?: string;
  mustSendCommand?: boolean;
}

interface IMatcher {
  (props: CommandState, ...args: any[]): Promise<ITestReport>;
}

type KeyOfMatcher = keyof typeof matchers;

function pickFn(name: KeyOfMatcher) {
  return matchers[name] as any as IMatcher;
}

function buildAndMatcherFunctions(params: ICreateMatcherParam) {
  const { commandName, isDebug, channelId, cordeBot, mustSendCommand } = params;

  const commandReturn: any = {};
  commandReturn.not = {};

  forEachPropsName(matchers, (matcher) => {
    commandReturn[matcher] = createMatcherFn({
      commandName,
      isDebug,
      isNot: false,
      matcher,
      channelId,
      cordeBot,
      mustSendCommand,
    });
    commandReturn.not[matcher] = createMatcherFn({
      commandName,
      isDebug,
      isNot: true,
      matcher,
      channelId,
      cordeBot,
      mustSendCommand,
    });
  });

  return commandReturn;
}

class CommandPromise extends Promise<any> implements corde.ICommandPromise {
  private readonly _param: ICreateMatcherParam;

  get and(): corde.AllCommandMatches {
    return buildAndMatcherFunctions({
      mustSendCommand: false,
      ...this._param,
    });
  }

  constructor(
    executor: (resolve: (value: any) => void, reject: (reason?: any) => void) => void,
    param: ICreateMatcherParam,
  ) {
    super(executor);
    this._param = param;
  }

  catch<TResult = never>(
    onrejected?: ((reason: any) => TResult | PromiseLike<TResult>) | null,
  ): CommandPromise {
    super.catch(onrejected);
    return this;
  }

  finally(onfinally?: (() => void) | null): CommandPromise {
    super.finally(onfinally);
    return this;
  }

  then<TResult1 = any, TResult2 = never>(
    onfulfilled?: ((value: any) => TResult1 | PromiseLike<TResult1>) | null,
    onrejected?: ((reason: any) => TResult2 | PromiseLike<TResult2>) | null,
  ): CommandPromise {
    super.then(onfulfilled, onrejected);
    return this;
  }

  [Symbol.toStringTag]: string;
}

async function resolveTestFunction(
  params: ICreateMatcherParam,
  fn: (...args: any[]) => Promise<ITestReport>,
): Promise<any> {
  try {
    const report = await fn();
    if (!report.pass && !params.isDebug) {
      report.trace = params.trace;
    }

    runtime.internalEvents.emit("test_end", report);

    if (params.isDebug) {
      return report;
    }
  } catch (error) {
    const failedReport: ITestReport = {
      pass: false,
      testName: params.matcher,
      message: handleError(error),
    };

    if (!params.isDebug) {
      failedReport.trace = params.trace;
    }

    runtime.internalEvents.emit("test_end", failedReport);
    if (params.isDebug) {
      return failedReport;
    }
  }
}

function createMatcherFn(params: ICreateMatcherParam) {
  const {
    matcher,
    isNot,
    commandName,
    channelId,
    isDebug,
    isCascade,
    cordeBot,
    guildId,
    mustSendCommand,
  } = params;
  const { testCollector, configs } = runtime;

  if (!testCollector.currentTestFile?.isInsideTestClosure && !isCascade) {
    throw new Error("command can only be used inside a test(it) closure");
  }

  return (
    ...args: any[]
  ): Promise<ITestReport | void | ((...args: any[]) => Promise<ITestReport>)> => {
    const trace = getStackTrace(undefined, true, matcher);
    // If someone pass expect.any, we must invoke it to return
    // the Any matcher.

    // If the suite is already marked as failed,
    // There is no need to run other tests.
    // Same for command assertion
    if (testCollector.currentSuite?.markedAsFailed) {
      return Promise.reject();
    }

    args = args.map((arg) => {
      if (arg === any) {
        return arg();
      }
      return arg;
    });

    const matcherFn = pickFn(matcher as KeyOfMatcher);

    let bot = runtime.bot;

    if (isDebug) {
      bot = cordeBot ?? runtime.bot;
    }

    const props = new CommandState({
      isNot,
      cordeBot: bot,
      command: commandName,
      timeout: configs.commandTimeout,
      guildId: guildId ?? configs.guildId,
      channelId: channelId ?? configs.channelId,
      testName: matcher,
      isCascade: isCascade ?? false,
      mustSendCommand: isNullOrUndefined(mustSendCommand) ? true : !!mustSendCommand,
    });

    const fn = matcherFn.bind(props, ...args);
    params.trace = trace;
    return new CommandPromise(async (resolve, reject) => {
      try {
        const result = await resolveTestFunction(params, fn);
        resolve(result);
      } catch (error) {
        reject(error);
      }
    }, params);
  };
}

function handleError(error: any) {
  if (error instanceof Error) {
    return error.message + "\n" + error.stack;
  }
  return error;
}

function createLocalCommand(isDebug: boolean) {
  let localCommand: any = {};

  localCommand = (commandName: any, channelId?: string, cordeBot?: ICordeBot) => {
    const commandReturn: any = {
      should: {
        not: {},
      },
    };

    forEachPropsName(matchers, (matcher) => {
      commandReturn.should[matcher] = createMatcherFn({
        commandName,
        isDebug,
        isNot: false,
        matcher,
        channelId,
        cordeBot,
      });
      commandReturn.should.not[matcher] = createMatcherFn({
        commandName,
        isDebug,
        isNot: true,
        matcher,
        channelId,
        cordeBot,
      });
    });

    return commandReturn;
  };

  localCommand.any = any;

  return localCommand;
}

export const command = createLocalCommand(false) as corde.ICommand;

type DebugTypes<T> = {
  [P in keyof T]: T[P] extends (...args: any[]) => Promise<any>
    ? (...params: Parameters<T[P]>) => Promise<{
        pass: boolean;
        testName: string;
        message: string;
        trace?: string;
      }>
    : T[P] extends Record<string, any>
    ? DebugTypes<T[P]>
    : T[P];
};

type MatchersWithNot = {
  not: typeof matchers;
} & typeof matchers;

type Matchers = {
  should: MatchersWithNot;
};

interface DebugFn {
  <T>(value: T, channelId?: string, cordeBot?: ICordeBot): DebugTypes<Matchers>;
}

/**
 * Testing object for corde
 * This is not used in production
 *
 * @internal
 */
export const debugCommand = createLocalCommand(true) as DebugFn;
