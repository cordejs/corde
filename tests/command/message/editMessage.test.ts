import MockDiscord from "../../mocks/mockDiscord";
import { createCordeBotWithMockedFunctions, testHelper } from "../../testHelper";
import { ICordeBot, ITestReport } from "../../../src/types";
import { MockEvents } from "../../mocks/mockEvents";
import { debugCommand } from "../../../src/command";
import { messageUtils } from "../../../src/command/matches/message/messageUtils";

const testName = "editMessage";

const failReport: ITestReport = {
  pass: false,
  testName,
};

const passReport: ITestReport = {
  pass: true,
  testName,
};

let mockDiscord = new MockDiscord();
let cordeClient = createCordeBotWithMockedFunctions(mockDiscord, mockDiscord.client);

function debugCon(customCommand?: string, customChannelId?: string, customClient?: ICordeBot) {
  return debugCommand(customCommand ?? "con", customChannelId, customClient ?? cordeClient);
}

describe(`testing ${testName} function`, () => {
  beforeEach(() => {
    [mockDiscord, cordeClient] = testHelper.initCommandTestsFixtures();
  });

  it("should return a failed test due to invalid parameter (null)", async () => {
    const report = await debugCon()
      // @ts-ignore
      .should.editMessage(undefined);
    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a failed test due to invalid parameter (undefined)", async () => {
    const report = await debugCon().should.editMessage(undefined);
    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should fail due to message was not edited by the bot (no messageIdentifier)", async () => {
    const report = await debugCon().should.editMessage("pong");
    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should fail due to message was not edited by the bot (no messageIdentifier as string)", async () => {
    const report = await debugCon().should.editMessage("pong", "123");
    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should fail due to message was not edited by the bot (no messageIdentifier as object with id)", async () => {
    const report = await debugCon().should.editMessage("pong", { id: "123" });
    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should fail due to message was not edited by the bot (no messageIdentifier as object with content)", async () => {
    const report = await debugCon().should.editMessage("pong", { oldContent: "message test" });
    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should get success test due to timeout but isNot = true", async () => {
    const report = await debugCon().should.not.editMessage("pong");
    expect(report).toEqual(passReport);
    expect(report).toMatchSnapshot();
  });

  it("should get success test due to bot returned equal message", async () => {
    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessageContentOrEmbedChange();

    const report = await debugCon().should.editMessage(mockDiscord.message.content);
    expect(report).toEqual(passReport);
    expect(report).toMatchSnapshot();
  });

  it("should get success test due to bot returned equal messages (string type)", async () => {
    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessageContentOrEmbedChange();

    const report = await debugCon().should.editMessage(mockDiscord.message.content);

    expect(report).toEqual(passReport);
    expect(report).toMatchSnapshot();
  });

  it("should get success test due to bot returned equal messages (string number)", async () => {
    mockDiscord.message.content = "2";

    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessageContentOrEmbedChange();

    const report = await debugCon().should.editMessage(2);

    expect(report).toEqual(passReport);
    expect(report).toMatchSnapshot();
  });

  it("should get success test due to bot returned equal messages (type embed)", async () => {
    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessageContentOrEmbedChange(mockDiscord.messageWithEmbed);

    const embedInterface = messageUtils.messageEmbedToMessageEmbedInterface(
      mockDiscord.messageEmbed,
    );
    const report = await debugCon().should.editMessage(embedInterface);

    expect(report).toEqual(passReport);
    expect(report).toMatchSnapshot();
  });

  it("should get failed test due to bot returned equal messages (isNot true)", async () => {
    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessageContentOrEmbedChange(mockDiscord.messageWithEmbed);

    const embedInterface = messageUtils.messageEmbedToMessageEmbedInterface(
      mockDiscord.messageEmbed,
    );
    const report = await debugCon().should.not.editMessage(embedInterface);

    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should get success test due to bot returned different messages (isNot true)", async () => {
    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessageContentOrEmbedChange();

    const report = await debugCon().should.not.editMessage(mockDiscord.messageEmbedSimple);

    expect(report).toEqual(passReport);
    expect(report).toMatchSnapshot();
  });

  it("should get fail test due to bot returned different messages (both embed)", async () => {
    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessageContentOrEmbedChange(mockDiscord.messageWithEmbed);

    const embedInternal: corde.IMessageEmbed = {
      fields: mockDiscord.messageEmbedSimple.fields,
    };

    const report = await debugCon().should.editMessage(embedInternal);

    expect(report).toMatchObject(failReport);
    // For some reason snapshot is falling in CI
  });

  it("should get fail test due to bot returned different messages (expect primitive and returned embed)", async () => {
    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessageContentOrEmbedChange(mockDiscord.messageWithEmbed);

    const expectValue = "expect value";

    const report = await debugCon().should.editMessage(expectValue);

    expect(report).toMatchObject(failReport);
    // For some reason snapshot is falling in CI
  });

  it("should get fail test due to bot returned different messages both primitive values", async () => {
    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessageContentOrEmbedChange();

    const expectValue = "expect value";

    const report = await debugCon().should.editMessage(expectValue);

    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should get fail test due to bot returned different messages (expect embed and returned primitive)", async () => {
    const mockEvent = new MockEvents(cordeClient, mockDiscord);
    mockEvent.mockOnceMessageContentOrEmbedChange();

    const events = new MockEvents(cordeClient, mockDiscord);
    events.mockOnceMessageContentOrEmbedChange();

    const report = await debugCon().should.editMessage(mockDiscord.messageEmbedSimple);

    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });

  it("should return a failed test due to failure in message sending", async () => {
    cordeClient.getRoles = jest.fn().mockReturnValue(mockDiscord.roleManager.cache);
    cordeClient.findRole = jest.fn().mockReturnValue(mockDiscord.role);

    const errorMessage = "can not send message to channel x";
    cordeClient.sendTextMessage = jest
      .fn()
      .mockImplementation(() => Promise.reject(new Error(errorMessage)));

    const report = await debugCon().should.editMessage("value", { id: "123" });

    expect(report).toMatchObject(failReport);
    expect(report).toMatchSnapshot();
  });
});
