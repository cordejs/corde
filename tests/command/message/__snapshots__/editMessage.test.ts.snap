// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing editMessage function should fail due to message was not edited by the bot (no messageIdentifier as object with content) 1`] = `
Object {
  "message": "    
    expected: testing bot to edit the message of content \\"message test\\"
    received: message was not edited
",
  "pass": false,
  "testName": "editMessage",
}
`;

exports[`testing editMessage function should fail due to message was not edited by the bot (no messageIdentifier as object with id) 1`] = `
Object {
  "message": "    
    expected: testing bot to edit the message of id 123
    received: message was not edited
",
  "pass": false,
  "testName": "editMessage",
}
`;

exports[`testing editMessage function should fail due to message was not edited by the bot (no messageIdentifier as string) 1`] = `
Object {
  "message": "    
    expected: testing bot to edit the message of id 123
    received: message was not edited
",
  "pass": false,
  "testName": "editMessage",
}
`;

exports[`testing editMessage function should fail due to message was not edited by the bot (no messageIdentifier) 1`] = `
Object {
  "message": "    
    expected: testing bot to edit the last message sent
    received: message was not edited
",
  "pass": false,
  "testName": "editMessage",
}
`;

exports[`testing editMessage function should get fail test due to bot returned different messages (both embed) 1`] = `
Object {
  "message": "    
    [32m- Expected[39m
    [31m+ Received[39m
    
    [2m  Object {[22m
    [32m-   \\"author\\": Object {[39m
    [32m-     \\"iconURL\\": \\"https://github.com/cordejs/corde/blob/master/website/static/img/logo/android-icon-192x192.png\\",[39m
    [32m-     \\"name\\": \\"cordeto\\",[39m
    [32m-     \\"url\\": undefined,[39m
    [32m-   },[39m
    [32m-   \\"color\\": 2123412,[39m
    [32m-   \\"description\\": \\"example description\\",[39m
    [2m    \\"fields\\": Array [[22m
    [2m      Object {[22m
    [2m        \\"inline\\": false,[22m
    [32m-       \\"name\\": \\"field1\\",[39m
    [32m-       \\"value\\": \\"1\\",[39m
    [31m+       \\"name\\": \\"Regular field title\\",[39m
    [31m+       \\"value\\": \\"Some value here\\",[39m
    [2m      },[22m
    [32m-   ],[39m
    [32m-   \\"footer\\": Object {[39m
    [32m-     \\"iconURL\\": \\"iconURl example\\",[39m
    [32m-     \\"proxyIconURL\\": \\"proxyiconURl example\\",[39m
    [32m-     \\"text\\": \\"test example\\",[39m
    [31m+     Object {[39m
    [31m+       \\"inline\\": false,[39m
    [31m+       \\"name\\": \\"â€‹\\",[39m
    [31m+       \\"value\\": \\"â€‹\\",[39m
    [2m      },[22m
    [32m-   \\"image\\": Object {[39m
    [32m-     \\"height\\": 100,[39m
    [32m-     \\"url\\": \\"url example\\",[39m
    [32m-     \\"width\\": 100,[39m
    [31m+     Object {[39m
    [31m+       \\"inline\\": true,[39m
    [31m+       \\"name\\": \\"Inline field title\\",[39m
    [31m+       \\"value\\": \\"Some value here\\",[39m
    [2m      },[22m
    [32m-   \\"thumbnailUrl\\": Object {[39m
    [32m-     \\"height\\": 100,[39m
    [32m-     \\"proxyURL\\": \\"proxy url example thumbnail\\",[39m
    [32m-     \\"url\\": \\"url example thumbnail\\",[39m
    [32m-     \\"width\\": 100,[39m
    [31m+     Object {[39m
    [31m+       \\"inline\\": true,[39m
    [31m+       \\"name\\": \\"Inline field title\\",[39m
    [31m+       \\"value\\": \\"Some value here\\",[39m
    [2m      },[22m
    [32m-   \\"timestamp\\": 886761031000,[39m
    [32m-   \\"title\\": \\"Some title\\",[39m
    [32m-   \\"url\\": \\"url example\\",[39m
    [31m+   ],[39m
    [2m  }[22m
",
  "pass": false,
  "testName": "editMessage",
}
`;

exports[`testing editMessage function should get fail test due to bot returned different messages (expect embed and returned primitive) 1`] = `
Object {
  "message": "    
    expected: Object {
      \\"author\\": Object {
        \\"iconURL\\": \\"https://i.imgur.com/wSTFkRM.png\\",
        \\"name\\": \\"Some name\\",
        \\"url\\": \\"https://discord.js.org\\",
      },
      \\"color\\": 2123412,
      \\"description\\": \\"Some description here\\",
      \\"fields\\": Array [
        Object {
          \\"inline\\": false,
          \\"name\\": \\"Regular field title\\",
          \\"value\\": \\"Some value here\\",
        },
        Object {
          \\"inline\\": false,
          \\"name\\": \\"â€‹\\",
          \\"value\\": \\"â€‹\\",
        },
        Object {
          \\"inline\\": true,
          \\"name\\": \\"Inline field title\\",
          \\"value\\": \\"Some value here\\",
        },
        Object {
          \\"inline\\": true,
          \\"name\\": \\"Inline field title\\",
          \\"value\\": \\"Some value here\\",
        },
      ],
      \\"thumbnailUrl\\": Object {
        \\"url\\": \\"https://i.imgur.com/wSTFkRM.png\\",
      },
      \\"title\\": \\"Some title\\",
      \\"url\\": \\"https://discord.js.org/\\",
    }
    received: 'this is the message content'
",
  "pass": false,
  "testName": "editMessage",
}
`;

exports[`testing editMessage function should get fail test due to bot returned different messages (expect primitive and returned embed) 1`] = `
Object {
  "message": "    
    expected: 'expect value'
    received: Object {
      \\"author\\": Object {
        \\"iconURL\\": \\"https://github.com/cordejs/corde/blob/master/website/static/img/logo/android-icon-192x192.png\\",
        \\"name\\": \\"cordeto\\",
        \\"url\\": undefined,
      },
      \\"color\\": 2123412,
      \\"description\\": \\"example description\\",
      \\"fields\\": Array [
        Object {
          \\"inline\\": false,
          \\"name\\": \\"field1\\",
          \\"value\\": \\"1\\",
        },
      ],
      \\"footer\\": Object {
        \\"iconURL\\": \\"iconURl example\\",
        \\"proxyIconURL\\": \\"proxyiconURl example\\",
        \\"text\\": \\"test example\\",
      },
      \\"image\\": Object {
        \\"height\\": 100,
        \\"url\\": \\"url example\\",
        \\"width\\": 100,
      },
      \\"thumbnailUrl\\": Object {
        \\"height\\": 100,
        \\"proxyURL\\": \\"proxy url example thumbnail\\",
        \\"url\\": \\"url example thumbnail\\",
        \\"width\\": 100,
      },
      \\"timestamp\\": 886761031000,
      \\"title\\": \\"Some title\\",
      \\"url\\": \\"url example\\",
    }
",
  "pass": false,
  "testName": "editMessage",
}
`;

exports[`testing editMessage function should get fail test due to bot returned different messages both primitive values 1`] = `
Object {
  "message": "    
    expected: 'expect value'
    received: 'this is the message content'
",
  "pass": false,
  "testName": "editMessage",
}
`;

exports[`testing editMessage function should get failed test due to bot returned equal messages (isNot true) 1`] = `
Object {
  "message": "    
    expected: message from bot be different from expectation
    received: both returned and expectation are equal
",
  "pass": false,
  "testName": "editMessage",
}
`;

exports[`testing editMessage function should get success test due to bot returned different messages (isNot true) 1`] = `
Object {
  "pass": true,
  "testName": "editMessage",
}
`;

exports[`testing editMessage function should get success test due to bot returned equal message 1`] = `
Object {
  "pass": true,
  "testName": "editMessage",
}
`;

exports[`testing editMessage function should get success test due to bot returned equal messages (string number) 1`] = `
Object {
  "pass": true,
  "testName": "editMessage",
}
`;

exports[`testing editMessage function should get success test due to bot returned equal messages (string type) 1`] = `
Object {
  "pass": true,
  "testName": "editMessage",
}
`;

exports[`testing editMessage function should get success test due to bot returned equal messages (type embed) 1`] = `
Object {
  "pass": true,
  "testName": "editMessage",
}
`;

exports[`testing editMessage function should get success test due to timeout but isNot = true 1`] = `
Object {
  "pass": true,
  "testName": "editMessage",
}
`;

exports[`testing editMessage function should return a failed test due to failure in message sending 1`] = `
Object {
  "message": "    
    can not send message to channel x
",
  "pass": false,
  "testName": "editMessage",
}
`;

exports[`testing editMessage function should return a failed test due to invalid parameter (null) 1`] = `
Object {
  "message": "    
    expected: expect value to be a primitive value (string, boolean, number) or an IMessageEmbed object
    received: undefined
",
  "pass": false,
  "testName": "editMessage",
}
`;

exports[`testing editMessage function should return a failed test due to invalid parameter (undefined) 1`] = `
Object {
  "message": "    
    expected: expect value to be a primitive value (string, boolean, number) or an IMessageEmbed object
    received: undefined
",
  "pass": false,
  "testName": "editMessage",
}
`;
